### DISCUSSION TITLE
Non-existing integrations / custom commands (shell, curl, etc)

### DISCUSSION THREAD

**Discussion by shafr** [mondoohq/cnquery#1232] [Q&A] [answered]
Hi there,

I'm looking for a way to query multiple resources **(onprem)**, such as:
* postgres
* kafka
* nagios
* haproxy
* httpd
* splunk

I haven't found existing integrations, so I was wondering if there is a way to run:
* linux shell command (which would run *-client & fetch required information)
* curl command with session support (for example first would be POST command to login, then to query)

In general what is the approach for non-existing integrations ?

**Comment by chris-rock** [comment, marked as answer]
@shafr Thank you for raising that question. You have quite some interesting potential integrations. I would love to hear a bit more about the use cases to give better guidance.

You always have the option to shell out.  With MQL you have the `command` resource available when you on a local target eg. `cnquery shell local`:

```
cnquery> data = command("curl http://date.jsontest.com/").stdout; parse.json(content: data).params
parse.json.params: {
  date: "05-23-2023"
  milliseconds_since_epoch: 1684866165523.000000
  time: "06:22:45 PM"
}
```

Then you could do assertions on top of that:

```
cnquery> data = command("curl http://date.jsontest.com/").stdout; parse.json(content: data).params["date"] == "05-23-2023"
[ok] value: "05-23-2023"
```

We also allow the addition of more providers and resource packs, as we've done https://github.com/mondoohq/cnquery/pull/1190 and https://github.com/mondoohq/cnquery/pull/1224. At this point we have not created the provider and resource pack documentation
because we plan to make providers and resource packs fully seperate plugins in the near future.

I hope that answers your question.

**Reply by shafr** [reply]
Yes, that is what I was looking for, thank you!
